micronaut-camunda-bpm-feature/src/main/java/info/novatec/micronaut/camunda/bpm/feature/metrics/CamundaMetrics.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package info.novatec.micronaut.camunda.bpm.feature.metrics;
+
+import io.micronaut.context.ApplicationContext;
+import org.camunda.bpm.engine.ManagementService;
+import org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl;
+import org.camunda.bpm.engine.management.MetricsQuery;
+
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.Map;
+
+abstract public class CamundaMetrics {
+
+    //@Property(name = MICRONAUT_METRICS_BINDERS + ".camunda.refreshWaitTime", defaultValue = "60")
+    private long refreshWaitTimeSeconds = 60;
+
+    private LocalDateTime lastUpdate = LocalDateTime.now();
+
+    private ManagementService managementService = null;
+
+    private final Map<String, MetricsQuery> metricsQueryMap = new HashMap<>();
+
+    abstract protected ApplicationContext getApplicationContext();
+
+    abstract protected void update();
+
+    protected void doUpdateIfNecessary() {
+        if (shouldLoad()) {
+            update();
+        }
+    }
+
+    protected boolean shouldLoad() {
+
+        LocalDateTime now = LocalDateTime.now();
+        if (now.isAfter(lastUpdate.minusMinutes(refreshWaitTimeSeconds))) {
+            lastUpdate = now;
+            return true;
+        }
+        return false;
+    }
+
+    protected long sumMetric(String metricName) {
+
+        MetricsQuery metricsQuery = getMetricsQuery(metricName);
+        if (metricsQuery != null) {
+            return metricsQuery.sum();
+        } else {
+            return 0;
+        }
+    }
+
+    private MetricsQuery getMetricsQuery(String metricsName) {
+        if (metricsQueryMap.containsKey(metricsName)) {
+            return metricsQueryMap.get(metricsName);
+        }
+        if (initManagementService()) {
+            MetricsQuery metricsQuery = managementService.createMetricsQuery().name(metricsName);
+            // does not return the intialized metrics query
+            metricsQueryMap.put(metricsName, metricsQuery);
+            return metricsQuery;
+        }
+        return null;
+    }
+
+    private boolean initManagementService() {
+        if (managementService == null) {
+            ProcessEngineConfigurationImpl processEngineConfiguration = getApplicationContext().getBean(ProcessEngineConfigurationImpl.class);
```

Comment:
Using application context here because injection of managementService or processEngine leads to infinity loop during initialization
metrics are init very early even before datasource is completly setup
but processengine needs datasource